staging/src/k8s.io/api/storage/v1beta1/types.go

```diff
@@ -209,3 +209,174 @@ type VolumeError struct {
+
+	// Specification of the CSI Driver.
+	Spec CSIDriverSpec `json:"spec" protobuf:"bytes,2,opt,name=spec"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// CSIDriverList is a collection of CSIDriver objects.
+type CSIDriverList struct {
+	metav1.TypeMeta `json:",inline"`
+
+	// Standard list metadata
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#metadata
+	// +optional
+	metav1.ListMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+
+	// items is the list of CSIDriver
+	Items []CSIDriver `json:"items" protobuf:"bytes,2,rep,name=items"`
+}
+
+// CSIDriverSpec is the specification of a CSIDriver.
+type CSIDriverSpec struct {
+	// attachRequired indicates this CSI volume driver requires an attach
+	// operation (because it implements the CSI ControllerPublishVolume()
+	// method), and that Kubernetes should call attach and wait for any attach
+	// operation to complete before proceeding to mounting.
+	// If value is not specified, default is false -- meaning attach will not be
+	// called.
+	// +optional
+	AttachRequired bool `json:"attachRequired,omitempty" protobuf:"varint,1,opt,name=attachRequired"`
+
+	// If specified, podInfoRequiredOnMount indicates this CSI volume driver
+	// requires additional pod information (like podName, podUID, etc.) during
+	// mount operations.
+	// If value is not specified, pod information will not be passed on mount.
+	// If value is set to a valid version, Kubelet will pass pod information as
+	// VolumeAttributes in the CSI NodePublishVolume() calls.
+	// Supported versions:
+	// Version "v1" will pass the following ValueAttributes
+	// "csi.storage.k8s.io/pod.name": pod.Name
+	// "csi.storage.k8s.io/pod.namespace": pod.Namespace
+	// "csi.storage.k8s.io/pod.uid": string(pod.UID)
+	// +optional
+	PodInfoOnMountVersion string `json:"podInfoOnMountVersion,omitempty" protobuf:"bytes,2,opt,name=podInfoOnMountVersion"`
+}
+
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// CSINodeInfo holds information about all CSI drivers installed on a node.
+type CSINodeInfo struct {
+	metav1.TypeMeta `json:",inline"`
+
+	// metadata.name must be the Kubernetes node name.
+	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+
+	// spec is the specification of CSINodeInfo
+	Spec CSINodeInfoSpec `json:"spec" protobuf:"bytes,2,opt,name=spec"`
+
+	// status is the current status of CSINodeInfo
+	Status CSINodeInfoStatus `json:"status" protobuf:"bytes,3,opt,name=status"`
+}
+
+// CSINodeInfoSpec holds information about the specification of all CSI drivers installed on a node
+type CSINodeInfoSpec struct {
+	// drivers is a list of specifications of CSIDriverInfo
+	// +patchMergeKey=name
+	// +patchStrategy=merge
+	Drivers []CSIDriverInfoSpec `json:"drivers" patchStrategy:"merge" patchMergeKey:"name" protobuf:"bytes,1,rep,name=drivers"`
```

Comment:
Can this be empty?

Reply:
Not opposed to the name, although I think `CSIDriverInfoSpec` came directly out of a prior API Review, storage did not come up with this name. It can be empty on initialization. We create the `CSINodeInfo` object first then populate it with `UPDATE` calls

Reply:
My concern with this name is that typicvally a "SomethingSpec" is directly related to a "Something".  In this case, tit is not related to a "CSIDriverInfo", which my brain then translates to "CSIDriver".  The only distinction is a meaningless "Info".  Calling it a "CSINodeDriverInfoSpec" keeps in my brain that it is related to a node-driver.  I might even call it "CSINodeDriver" and the status "CSINodeDriverStatus", just to break the "spec" pattern.

Ths is a relatively minor point.  If search-and-replace can't solve it, it may not be worth the effort